File: 116-PopulatingNextRightPointers/improveRecursive.java
class Solution {
    public Node connect(Node root) {
        dfs(root, null);
        return root;
    }

    private void dfs(Node node, Node next) {
        if(node != null) {
            node.next = next;
            dfs(node.left, node.right);
            dfs(node.right, node.next != null ? node.next.left : null);
        }
    }
};

class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}
    
    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};

File: 235-LowestCommonAncestorOfBST/improveRecursive.cpp
#include <iostream>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        int temp = root->val;
        int left = p->val;
        int right = q->val;
        //利用了BST的性质
        if (left > temp && right > temp) {
            return lowestCommonAncestor(root->right, p, q);
        }
        else if (left < temp && right < temp) {
            return lowestCommonAncestor(root->left, p, q);
        }
        else return root;
    }
};